src/mount/mount-tool.c

```diff
@@ -680,28 +691,191 @@ static int stop_mount(
+                if (r < 0) {
+                        if (r2 >= 0)
+                                r2 = r;
+                        continue;
+                }
+
+                r = stop_mounts(bus, where, false);
+                if (r < 0) {
+                        if (r2 >= 0)
+                                r2 = r;
+                        continue;
+                }
+        }
+
+        return r;
+}
+
+static int umount_by_device(sd_bus *bus, const char *what) {
+        _cleanup_udev_device_unref_ struct udev_device *d = NULL;
+        _cleanup_udev_unref_ struct udev *udev = NULL;
+        _cleanup_free_ char *where = NULL;
+        struct stat st;
+        const char *v;
+        int r = 0, r2 = 0;
+
+        assert(what);
+
+        if (stat(what, &st) < 0)
+                return log_error_errno(errno, "Can't stat %s: %m", what);
+
+        if (!S_ISBLK(st.st_mode)) {
+                log_error("Path %s is not a block device, don't know what to do.", what);
+                return -ENOTBLK;
+        }
+
+        udev = udev_new();
+        if (!udev)
+                return log_oom();
+
+        d = udev_device_new_from_devnum(udev, 'b', st.st_rdev);
+        if (!d)
+                return log_oom();
+
+        v = udev_device_get_property_value(d, "ID_FS_USAGE");
+        if (!streq_ptr(v, "filesystem")) {
+                log_error("%s does not contain a file system.", what);
+                return -EINVAL;
+        }
+
+        v = udev_device_get_property_value(d, "SYSTEMD_MOUNT_WHERE");
+        if (!isempty(v)) {
+                where = strdup(v);
+                r = stop_mounts(bus, where, false);
+        }
+
+        v = udev_device_get_devnode(d);
+        if (!isempty(v))
+                r2 = umount_by_mountinfo(bus, v);
+
+        if (r < 0)
+                return r;
+
+        return r2;
+}
+
+static int action_umount(
+                sd_bus *bus,
+                int argc,
+                char **argv) {
+
+        int i, r, r2 = 0;
+
+        for (i = optind; i < argc; i++) {
+                _cleanup_free_ char *what = NULL, *where = NULL, *u = NULL;
+
+                u = fstab_node_to_udev_node(argv[i]);
+                if (!strcmp(u, argv[i])) {
+                        r = path_make_absolute_cwd(argv[i], &where);
+                        if (r < 0) {
+                                log_error_errno(r, "Failed to make path absolute: %m");
+                                if (r2 >= 0)
+                                        r2 = r;
+                        } else {
+                                r = stop_mounts(bus, where, is_device_path(where));
+                                if (r >= 0)
+                                        continue;
+                                if (r < 0 && r2 >= 0)
+                                        r2 = r;
+                        }
+                }
+
+                if (arg_transport != BUS_TRANSPORT_LOCAL)
```

Comment:
this check appears, misplaced. if we are not local, then the path_make_absolute_cwd() above can't work either